### DISCUSSION TITLE
🎷JAX at NeurIPS 2020 - JAX on Cloud TPUs, JAX at DeepMind, JAX MD 🎷

### DISCUSSION THREAD

**Discussion by 8bitmp3** [jax-ml/jax#5027] [General] [answered]
Hey core dev team @mattjj @skye @dougalm @hawkinsp @jekbradbury @gnecula @apaszke @jakevdp @zhangqiaorjc (did I miss anyone, @ Chris Leary?) 👋  

Do you have any plans to present at NeurIPS or any other upcoming big ML conference any time soon? I noticed the Program Transformations for ML isn't in the NeurIPS workshop schedule this year https://github.com/program-transformations/program-transformations.github.io Maybe the workshop is on hold this year 🤔 cc @dynamicwebpaige 

(I'm aware of a few recent JAX talks like https://www.youtube.com/watch?v=z-WSrQDXkuM by @jakevdp 
and https://www.youtube.com/watch?v=lN3wWUdB_UM + https://www.youtube.com/watch?v=BzuEGdGHKjc by @mattjj and also this one https://www.youtube.com/watch?v=mbUwCPiqZBM by @froystig)

<img width="500" alt="image" src="https://user-images.githubusercontent.com/19637339/100474383-6164d780-30d8-11eb-9c8e-ab246f8d262c.png">

**Comment by mattjj** [comment]
Thanks for the question!

There may be a NeurIPS demo planned (cc @jekbradbury @skye).

I'm co-presenting a tutorial on deep implicit layers. That's not on JAX, but there is some JAX code involved!

**Reply by 8bitmp3** [reply]
Thanks for your reply 👋 Which session is the deep implicit layer talk @mattjj?

![](https://nkayesel.files.wordpress.com/2013/04/537.gif)

**Reply by mattjj** [reply]
It's a [tutorial](https://neurips.cc/Conferences/2020/Schedule?type=Tutorial), so it's [on Monday 12/07](https://neurips.cc/virtual/2020/public/cal_main.html).

**Reply by 8bitmp3** [reply]
@mattjj great post by DeepMind today (repost from @dynamicwebpaige):

> We have found that JAX has enabled rapid experimentation with novel algorithms and architectures and it now underpins many of our recent publications. To learn more please consider joining our JAX Roundtable, Wednesday December 9th 7:00pm GMT, at the NeurIPS virtual conference.

https://deepmind.com/blog/article/using-jax-to-accelerate-our-research

<img width="100" alt="image" src="https://user-images.githubusercontent.com/19637339/101206790-8a9cdf00-3667-11eb-8779-162a8a2f1c07.png">

**Reply by 8bitmp3** [reply]
To sum up, here's the JAX-specific stuff at NeurIPS so far🔥:

- December 8: **JAX 💟 Cloud TPUs** (11am-12pm PT/7pm-8pm GMT):
    - Expo link: `https://neurips.cc/accounts/login/?next=/ExpoConferences/2020/Sponsorpage%3Fid%3D656` (make sure you're registered first)
- December 9: **DeepMind 💙 JAX** (Haiku, Optax, RLax, Jraph, etc) (11am-12pm PST/7pm-9pm GMT):
    - Blog post: `https://deepmind.com/blog/article/using-jax-to-accelerate-our-research`
- December 9/10: **JAX MD: A Framework for Differentiable Physics** (Orals & Spotlights Track 23: Graph/Meta Learning/Software, Poster Session 5) https://neurips.cc/virtual/2020/public/poster_83d3d4b6c9579515e1679aca8cbc8033.html / [paper](https://proceedings.neurips.cc/paper/2020/file/83d3d4b6c9579515e1679aca8cbc8033-Paper.pdf)

- December 9: **JAX team Q&A** (9:30am-10am PST/5.30pm-6pm GMT)
- December 10 (11): **JAX team Q&A 🗣** (6pm-6.30pm PST/2am-2.30am GMT)
also:

- December 7: **Deep Implicit Layers: Neural ODEs, Equilibrium Models, and Differentiable Optimization** with @mattjj (1.30pm-4pm PT, Tutorial Track 3C) `https://neurips.cc/Conferences/2020/Schedule?showEvent=16653` (JAX Colabs, YouTube 📺 recording, etc http://implicit-layers-tutorial.org)

<img width="477" alt="image" src="https://user-images.githubusercontent.com/19637339/101226715-4ae8ee00-368d-11eb-9d60-ae05b5adbdaa.png">

(updated with Q&As from @jekbradbury)

**Reply by jekbradbury** [reply]
Yeah, that's just about everything! We also have two other Q&As planned with folks from the JAX core team (in addition to the Q&As attached to the talks you've listed):

Wednesday 9:30-10AM PST/17:30-18:00 GMT (Europe-friendly time)
Thursday 6-6:30PM PST/Friday 02:00-02:30 GMT (Asia-friendly time)

The Meet links will be added to the same agenda page as me and Skye's talk.

**Reply by 8bitmp3** [reply]
@mattjj Monday, December 7th's session - _Neural ODEs, Equilibrium Models and Beyond_ - is "video 5" on the tutorial page. Are we allowed to share the SlidesLive link here?

[Update]

https://slideslive.com/38935810/deep-implicit-layers-neural-odes-equilibrium-models-and-beyond

**Colabs**:

– Introduction https://colab.research.google.com/drive/1OUVzeUh66wVOFI_Nc_rIAuO70gHimHH8?usp=sharing
– Implicit functions and automatic differentiation https://colab.research.google.com/drive/17C9age7vi9DpUIfs_sFBWtDb4DAylbVe?usp=sharing
– Neural ordinary differential equations https://colab.research.google.com/drive/1o2jWvmrQYjX99hZpTF415uT1rc35Mvft?usp=sharing
– Deep equilibrium models https://colab.research.google.com/drive/1yQd4j776iqbWhBDiObaDn002NsQRa3BK?usp=sharing
– Differentiable optimization https://colab.research.google.com/drive/1OWaxmU8JhNhy8atO2sH5BB0AcCpYCw0W?usp=sharing

Full details: http://implicit-layers-tutorial.org

**Reply by mattjj** [reply]
Sure! (Though we're still finalizing the captions and the notes webpage, so it might be best if folks watch it at or after the time for which it's scheduled.)

**Reply by 8bitmp3** [reply]
Professor Matt @mattjj, the students are appreciating your explanations of the implicit function theorem, JVPs and reverse-mode VJPs:

<img width="362" alt="image" src="https://user-images.githubusercontent.com/19637339/101412406-7ad90100-38da-11eb-84c5-006bcdb27a66.png">

---

> we're still finalizing the captions and the notes webpage

I hope they implement the changes soon, though I don't mind calculating "radiance" with respect to variables:

<img width="354" alt="image" src="https://user-images.githubusercontent.com/19637339/101412637-d73c2080-38da-11eb-8e05-f84710890629.png">

<img width="520" alt="image" src="https://user-images.githubusercontent.com/19637339/101412922-52053b80-38db-11eb-8d34-08baee161f6a.png">

[Update]

<img width="308" alt="image" src="https://user-images.githubusercontent.com/19637339/101413844-e91ec300-38dc-11eb-9c9a-38e394beaebc.png">

---


[Update]

From the chat - that's probably for @apaszke re: DEX:

<img width="325" alt="image" src="https://user-images.githubusercontent.com/19637339/101418315-7cf48d00-38e5-11eb-917f-63913a3857d5.png">

**Reply by jekbradbury** [reply]
Wait, @8bitmp3, where are these chats from?

**Reply by 8bitmp3** [reply]
@jekbradbury Rocketchat plugin that the organizers use for non-Zoom conversations -  https://neurips.cc/virtual/2020/protected/tutorial_ad1df793247a0e650d0d7166341b8d97.html (login required). There was even a comment from Yoshua.

**Reply by mattjj** [reply]
Thanks, @8bitmp3 !

We spent a while polishing the captions, but somehow they didn't come through to the stream. The "on-demand" video has better (but less entertaining) captions.

**Reply by 8bitmp3** [reply]
@jekbradbury 

There was a comment on deep equilibrium models/DEQs' time to compute, in case you can't find it:

Q: If DEQs match the feedforward models' parameter counts, what's the compute time like?
A: For similar architectures/param numbers, "usually 2x-3x longer" because "the number of forward iterations we often find a suitably accurate fixed point is more than the "typical" depth of many networks"... but improving is "a direction of ongoing work"... "To clarify:... 1.5x at inference" (approx.).

Yoshua also said earlier that "a multi-layer DEQ may still help for the same computation resources. The same argument has been made for deep RNNs."

**Reply by jekbradbury** [reply]
We opened up a general JAX discussion channel on rocket.chat at #google-research-jax, in case anyone is watching this thread

**Comment by 8bitmp3** [comment, marked as answer]
Reposting for visibility, since GitHub Discussions don't unroll some conversations automatically.

---

To sum up, here's the JAX-specific stuff at NeurIPS that you can find ~~so far~~ after the conference (UPDATED):

- December 8: **JAX 💟 Cloud TPUs** (11am-12pm PT/7pm-8pm GMT):

    - UPDATE: if you missed it - check the slides and 📺 https://twitter.com/jekbradbury/status/1337528357517291520 by @jekbradbury 

- December 9: **DeepMind 💙 JAX** (Haiku, Optax, RLax, Jraph, etc) (11am-12pm PST/7pm-9pm GMT): (https://twitter.com/DynamicWebPaige/status/1336735477928345601 - by @dynamicwebpaige)
    - Blog post: `https://deepmind.com/blog/article/using-jax-to-accelerate-our-research`

- December 9/10: **JAX MD: A Framework for Differentiable Physics** (Orals & Spotlights Track 23: Graph/Meta Learning/Software, Poster Session 5) https://neurips.cc/virtual/2020/public/poster_83d3d4b6c9579515e1679aca8cbc8033.html / [paper](https://proceedings.neurips.cc/paper/2020/file/83d3d4b6c9579515e1679aca8cbc8033-Paper.pdf)

also:

- December 7: **Deep Implicit Layers: Neural ODEs, Equilibrium Models, and Differentiable Optimization** with @mattjj (1.30pm-4pm PT, Tutorial Track 3C) `https://neurips.cc/Conferences/2020/Schedule?showEvent=16653`
  - JAX Colabs, YouTube 📺 recording, and more at http://implicit-layers-tutorial.org

(updated with Q&As from @jekbradbury, DeepMind talk public details from @dynamicwebpaige)

<img width="500" alt="image" src="https://user-images.githubusercontent.com/19637339/101555756-2d7b9300-39b1-11eb-80aa-5b7f1a70925a.png">

<img width="300" alt="image" src="https://user-images.githubusercontent.com/19637339/101555827-50a64280-39b1-11eb-85a3-a0a55f10c160.png">
